trin-core/src/types/validation.rs

```diff
@@ -0,0 +1,103 @@
+use anyhow::anyhow;
+use async_trait::async_trait;
+use serde_json::{json, Value};
+
+use crate::{
+    jsonrpc::{
+        service::dispatch_infura_request,
+        types::{HistoryJsonRpcRequest, JsonRequest, Params},
+    },
+    portalnet::types::{content_key::IdentityContentKey, messages::ByteList},
+    utils::infura::INFURA_BASE_URL,
+};
+
+/// Responsible for dispatching cross-overlay-network requests
+/// for data to perform validation. Currently, it just proxies these requests
+/// on to infura.
+#[derive(Debug, Clone)]
+pub struct HeaderOracle {
+    pub infura_url: String,
+    // We could simply store the main portal jsonrpc tx channel here, rather than each
+    // individual channel. But my sense is that this will be more useful in terms of
+    // determining which subnetworks are actually available.
+    pub history_jsonrpc_tx: Option<tokio::sync::mpsc::UnboundedSender<HistoryJsonRpcRequest>>,
+    pub header_gossip_jsonrpc_tx: Option<bool>,
+    pub block_indices_jsonrpc_tx: Option<bool>,
+    pub state_jsonrpc_tx: Option<bool>,
+}
+
+impl Default for HeaderOracle {
+    fn default() -> Self {
+        Self {
+            infura_url: INFURA_BASE_URL.to_string(),
+            history_jsonrpc_tx: None,
+            header_gossip_jsonrpc_tx: None,
+            block_indices_jsonrpc_tx: None,
+            state_jsonrpc_tx: None,
+        }
+    }
+}
+
+impl HeaderOracle {
+    // Currently falls back to infura, to be updated to use canonical block indices network.
+    pub fn get_hash_at_height(&mut self, hex_number: String) -> anyhow::Result<String> {
```

Comment:
Passing in the block number as a hex string at this level of the API is surprising to me. I guess it's because Infura/json-rpc expects a hex string, but that seems like an implementation detail that's leaking up into the API. I would expect a uint here, then converting to a hex string inside of `get_hash_at_height()` before building the `JsonRequest`.